### PR TITLE
feat (core\conf \log\db): item.dbc, sItemStore, item_dbc, item enforcement conf, subclass fix

### PR DESCRIPTION
**Pull request by acidmanifesto**
## Changes Proposed:
-  Adds item.dbc into the checks and actual use the item.dbc, we currently prior to this pr dont even load it up in the dbc store.
-  Adds Item DBC enforcement conf, Disallow overriding item attributes stored in DBC files with values from the database
- Reorder ifs for dbc enforcement
- add item_dbc support table for overrides of items via world db table instead of dbc edits.
- add item.dbc reading from actual item.dbc
- sItemStore added
- replace item template lookup with itemstore dbc item lookup.
- add additional logs
- changed rehash to reserve because of bare basic definitions:
rehash - takes an existing map and rebuilds a new size of buckets, rehashing in the process and redistributing elements into the new buckets.
reserve- guarantees you that if you don't insert more than the reserved number of elements, there will be no rehashing (i.e. your iterators will remain valid).
- Subclass matching showed a error on my local and workflow so sql patch to fix mixmatched subclass itemtemplate and item dbc enforcement is included
- new conf:
```

#
#   DBC.EnforceItemAttributes
#        Disallow overriding item attributes stored in DBC files with values from the database
#        Default: 0 - Off, Use DB values
#                 1 - On, Enforce DBC Values (default)

DBC.EnforceItemAttributes = 1

```
TC-CP and further inspiration:
https://github.com/TrinityCore/TrinityCore/commit/0c44bd33ee7b42c924859139a9f4b04cf2b91261
https://github.com/TrinityCore/TrinityCore/commit/fd26c3c87c5e75952d5af033a66c9a28c9af29b9

## Issues Addressed:

- Closes No item.dbc support
- Closes Subclass and Item DBC enforcement mismatch
image

## Tests Performed:

- builds and runs vs2022 on win 11
- self tested

## How to Test the Changes:

A Few Ways to test:

Conf DBC.EnforceItemAttributes = 1

1. Run on world: Custom item by @menevia16a
```
DELETE FROM `item_template` WHERE `entry`=57145;
INSERT INTO `item_template` (`entry`, `class`, `subclass`, `SoundOverrideSubclass`, `name`, `displayid`, `Quality`, `Flags`, `FlagsExtra`, `BuyCount`, `BuyPrice`, `SellPrice`, `InventoryType`, `AllowableClass`, `AllowableRace`, `ItemLevel`, `RequiredLevel`, `RequiredSkill`, `RequiredSkillRank`, `requiredspell`, `requiredhonorrank`, `RequiredCityRank`, `RequiredReputationFaction`, `RequiredReputationRank`, `maxcount`, `stackable`, `ContainerSlots`, `StatsCount`, `stat_type1`, `stat_value1`, `stat_type2`, `stat_value2`, `stat_type3`, `stat_value3`, `stat_type4`, `stat_value4`, `stat_type5`, `stat_value5`, `stat_type6`, `stat_value6`, `stat_type7`, `stat_value7`, `stat_type8`, `stat_value8`, `stat_type9`, `stat_value9`, `stat_type10`, `stat_value10`, `ScalingStatDistribution`, `ScalingStatValue`, `dmg_min1`, `dmg_max1`, `dmg_type1`, `dmg_min2`, `dmg_max2`, `dmg_type2`, `armor`, `holy_res`, `fire_res`, `nature_res`, `frost_res`, `shadow_res`, `arcane_res`, `delay`, `ammo_type`, `RangedModRange`, `spellid_1`, `spelltrigger_1`, `spellcharges_1`, `spellppmRate_1`, `spellcooldown_1`, `spellcategory_1`, `spellcategorycooldown_1`, `spellid_2`, `spelltrigger_2`, `spellcharges_2`, `spellppmRate_2`, `spellcooldown_2`, `spellcategory_2`, `spellcategorycooldown_2`, `spellid_3`, `spelltrigger_3`, `spellcharges_3`, `spellppmRate_3`, `spellcooldown_3`, `spellcategory_3`, `spellcategorycooldown_3`, `spellid_4`, `spelltrigger_4`, `spellcharges_4`, `spellppmRate_4`, `spellcooldown_4`, `spellcategory_4`, `spellcategorycooldown_4`, `spellid_5`, `spelltrigger_5`, `spellcharges_5`, `spellppmRate_5`, `spellcooldown_5`, `spellcategory_5`, `spellcategorycooldown_5`, `bonding`, `description`, `PageText`, `LanguageID`, `PageMaterial`, `startquest`, `lockid`, `Material`, `sheath`, `RandomProperty`, `RandomSuffix`, `block`, `itemset`, `MaxDurability`, `area`, `Map`, `BagFamily`, `TotemCategory`, `socketColor_1`, `socketContent_1`, `socketColor_2`, `socketContent_2`, `socketColor_3`, `socketContent_3`, `socketBonus`, `GemProperties`, `RequiredDisenchantSkill`, `ArmorDamageModifier`, `duration`, `ItemLimitCategory`, `HolidayId`, `ScriptName`, `DisenchantID`, `FoodType`, `minMoneyLoot`, `maxMoneyLoot`, `flagsCustom`, `VerifiedBuild`) VALUES 
(57145, 2, 7, -1, 'MDic\'s Long John', 41179, 4, 0, 0, 1, 0, 0, 13, -1, -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 7, 69, 3, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 333, 666, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2400, 0, 0, 61079, 2, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 'MDic\'s Long John with a built in curve to really hit that GSpot. Now equipped with nuclear powered vibrations. ALL PRAISE MDIC!', 0, 0, 0, 0, 0, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, '', 0, 0, 0, 0, 0, NULL);

DELETE FROM `item_dbc` WHERE `ID`=57145;
INSERT INTO `item_dbc` (`ID`, `ClassID`, `SubclassID`, `Sound_Override_Subclassid`, `Material`, `DisplayInfoID`, `InventoryType`, `SheatheType`) VALUES 
(57145, 2, 7, -1, 1, 41179, 13, 3);
```
2. log in with any sword user
3. .additem 57145
4. observe item is now in use and had displayimage when equiped

Log testing:
1. run on the world db:
```
UPDATE `item_dbc` SET `SheatheType`='0' WHERE  `ID`=57145;
```
2. restart world.
3. observe new error log stating item has wrong dbc value.

conf:
Conf DBC.EnforceItemAttributes = 0

1. restart world.
2. observe error in world console still.
3. dbc enforcement is not in play for item.

At the end of the test you can run on the world if you wish to correct and keep the custom test item:
```
UPDATE `item_dbc` SET `SheatheType`='3' WHERE  `ID`=57145;
```
Or run  to remove the custom item.
```
DELETE FROM `item_template` WHERE `entry`=57145;
DELETE FROM `item_dbc` WHERE `ID`=57145;
```

## Additional Item for fun
Run this on the world for a tome that will allow a special flying mount on non flight zone areas. This was a custom item from stygiancore days.
```
DELETE FROM `item_template` WHERE `entry`=701000;
INSERT INTO `item_template` (`entry`, `class`, `subclass`, `SoundOverrideSubclass`, `name`, `displayid`, `Quality`, `Flags`, `FlagsExtra`, `BuyCount`, `BuyPrice`, `SellPrice`, `InventoryType`, `AllowableClass`, `AllowableRace`, `ItemLevel`, `RequiredLevel`, `RequiredSkill`, `RequiredSkillRank`, `requiredspell`, `requiredhonorrank`, `RequiredCityRank`, `RequiredReputationFaction`, `RequiredReputationRank`, `maxcount`, `stackable`, `ContainerSlots`, `StatsCount`, `stat_type1`, `stat_value1`, `stat_type2`, `stat_value2`, `stat_type3`, `stat_value3`, `stat_type4`, `stat_value4`, `stat_type5`, `stat_value5`, `stat_type6`, `stat_value6`, `stat_type7`, `stat_value7`, `stat_type8`, `stat_value8`, `stat_type9`, `stat_value9`, `stat_type10`, `stat_value10`, `ScalingStatDistribution`, `ScalingStatValue`, `dmg_min1`, `dmg_max1`, `dmg_type1`, `dmg_min2`, `dmg_max2`, `dmg_type2`, `armor`, `holy_res`, `fire_res`, `nature_res`, `frost_res`, `shadow_res`, `arcane_res`, `delay`, `ammo_type`, `RangedModRange`, `spellid_1`, `spelltrigger_1`, `spellcharges_1`, `spellppmRate_1`, `spellcooldown_1`, `spellcategory_1`, `spellcategorycooldown_1`, `spellid_2`, `spelltrigger_2`, `spellcharges_2`, `spellppmRate_2`, `spellcooldown_2`, `spellcategory_2`, `spellcategorycooldown_2`, `spellid_3`, `spelltrigger_3`, `spellcharges_3`, `spellppmRate_3`, `spellcooldown_3`, `spellcategory_3`, `spellcategorycooldown_3`, `spellid_4`, `spelltrigger_4`, `spellcharges_4`, `spellppmRate_4`, `spellcooldown_4`, `spellcategory_4`, `spellcategorycooldown_4`, `spellid_5`, `spelltrigger_5`, `spellcharges_5`, `spellppmRate_5`, `spellcooldown_5`, `spellcategory_5`, `spellcategorycooldown_5`, `bonding`, `description`, `PageText`, `LanguageID`, `PageMaterial`, `startquest`, `lockid`, `Material`, `sheath`, `RandomProperty`, `RandomSuffix`, `block`, `itemset`, `MaxDurability`, `area`, `Map`, `BagFamily`, `TotemCategory`, `socketColor_1`, `socketContent_1`, `socketColor_2`, `socketContent_2`, `socketColor_3`, `socketContent_3`, `socketBonus`, `GemProperties`, `RequiredDisenchantSkill`, `ArmorDamageModifier`, `duration`, `ItemLimitCategory`, `HolidayId`, `ScriptName`, `DisenchantID`, `FoodType`, `minMoneyLoot`, `maxMoneyLoot`, `flagsCustom`, `VerifiedBuild`) VALUES (701000, 9, 0, -1, 'Tome of World Flying', 61330, 7, 134217792, 0, 1, 4500000, 4500000, 0, -1, -1, 80, 45, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1000, 0, 0, 483, 0, -1, 0, -1, 0, -1, 31700, 6, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 0, 0, 0, -1, 0, -1, 0, 'Learn to fly everywhere', 0, 0, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0, '', 0, 0, 0, 0, 0, 1);
```
## Known Issues and TODO List:

- [ ] At the moment we get no displayicon in game when dbc enforcement is off if the item is banked, issue unqiue to azerothcore for some reason for some items.
- [ ] At the moment we get a ? displayicon in game when dbc enforcement is off if the item is in the inventory issue unqiue to azerothcore for some reason for some items.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -2636,8 +2636,10 @@ void ObjectMgr::LoadItemTemplates()
         return;
     }
 
-    _itemTemplateStore.rehash(result->GetRowCount());
+    _itemTemplateStore.reserve(result->GetRowCount());
```
Comment (acidmanifesto) [inline src/server/game/Globals/ObjectMgr.cpp:2639]: before i got people wanting to bitch and complain about this without any or no explaination besides "why?". i advise you to look up what they first, and the reason why i changed it was due to it crashing.

Comment (Kitzunu) [reply src/server/game/Globals/ObjectMgr.cpp:2639]: Why? 💋

Comment (acidmanifesto) [reply src/server/game/Globals/ObjectMgr.cpp:2639]: 🖕

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 2]
```diff
@@ -1307,6 +1307,14 @@ DeletedCharacterTicketTrace = 0
 
 DungeonFinder.OptionsMask = 5
 
+#
+#   DBC.EnforceItemAttributes
+#        Disallow overriding item attributes stored in DBC files with values from the database
+#        Default: 0 - Off, Use DB values
+#                 1 - On, Enforce DBC Values (default)
+
+DBC.EnforceItemAttributes = 1
```
Comment (heyitsbench) [inline src/server/apps/worldserver/worldserver.conf.dist:1316]: Is this config really necessary? AC already overrides all other DBC tables from DB values with no config option for specific tables, what is it about item.DBC that necessitates this?

Comment (acidmanifesto) [reply src/server/apps/worldserver/worldserver.conf.dist:1316]: > 

the conf gives us the option to enforce the dbc value or not and to use the db value.
at the moment we do not using the dbc at all as evident with me adding it in the dbcstore. 
there is no over ride table for item_dbc at all currently as seen in the dbcstore.
we have alot of hard coded defines values being used and not using the dbc at all as evident in the objectmgr.cpp.
why we are not using the item.dbc at all is beyond me.

File: data/sql/updates/pending_db_world/Item.sql [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+UPDATE `item_template` SET `subclass`='4' WHERE  `entry`=17;
+UPDATE `item_template` SET `subclass`='6' WHERE  `entry`=2556;
+UPDATE `item_template` SET `subclass`='0' WHERE  `entry`=20221;
+UPDATE `item_template` SET `subclass`='13' WHERE  `entry`=31802;
+UPDATE `item_template` SET `subclass`='3' WHERE  `entry`=33080;
+UPDATE `item_template` SET `subclass`='3' WHERE  `entry`=33604;
+UPDATE `item_template` SET `subclass`='8' WHERE  `entry`=37445;
+UPDATE `item_template` SET `subclass`='12' WHERE  `entry`=37677;
+UPDATE `item_template` SET `subclass`='7' WHERE  `entry`=41749;
+UPDATE `item_template` SET `subclass`='1' WHERE  `entry`=53048;
```
Comment (heyitsbench) [inline data/sql/updates/pending_db_world/Item.sql:15]: (`subclass`=`'Integer'`)? :3

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/Item.sql:15]: also dont use quotes on ints :D

Comment (heyitsbench) [reply data/sql/updates/pending_db_world/Item.sql:15]: > also dont use quotes on ints :D

Literally what I said. >:(

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/Item.sql:15]: Me no speak french

Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/Item.sql:15]: Atm im pretty blitz. will deal with it tomorrow when the tv stops screaming at me.

Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/Item.sql]: back ticks are now dead.

Comment (Kitzunu) [reply data/sql/updates/pending_db_world/Item.sql]: and so am I... Inside :P

**Review by IntelligentQuantum — APPROVED** [review]

**Comment by acidmanifesto** [comment]
i moved some checks from item templates to dbc lookup with the sItemStore implemenation based on tc cherrypick and core comparisons. Should help it run more efficient then constant hits on to the item templates in the db as we are now using the item.dbc now with this pr.

File: src/server/game/DataStores/DBCStores.cpp [hunk 4]
```diff
@@ -97,7 +97,7 @@ DBCStorage <GtRegenHPPerSptEntry>         sGtRegenHPPerSptStore(GtRegenHPPerSptf
 DBCStorage <GtRegenMPPerSptEntry>         sGtRegenMPPerSptStore(GtRegenMPPerSptfmt);
 
 DBCStorage <HolidaysEntry>                sHolidaysStore(Holidaysfmt);
-
+DBCStorage <ItemEntry>                    sItemStore(Itemfmt);
```
Comment (Kitzunu) [inline src/server/game/DataStores/DBCStores.cpp:100]: add a blank line above

**Review by Kitzunu — CHANGES_REQUESTED** [review]
Will this essentially make item_template an override template from item.dbc?

File: src/server/game/DataStores/DBCStores.cpp (hunk 4 above)
Comment (acidmanifesto) [reply src/server/game/DataStores/DBCStores.cpp:100]: Gonna have to wait a week or two.

**Comment by acidmanifesto** [comment]
> Will this essentially make item_template an override template from item.dbc?

No. Item_dbc is the over ride for item.dbc that we are not even using. Select item template dbc info is in item the item dbc. We had some modified item templates not matching their select properties in the dbc, so the sql as u see corrects it with the error logging i created above in the issue template.

File: src/server/game/DataStores/DBCStores.cpp (hunk 4 above)
Comment (acidmanifesto) [reply src/server/game/DataStores/DBCStores.cpp:101]: https://github.com/azerothcore/azerothcore-wotlk/pull/14675/commits/8159dbc4edc680e252c4ef85847eb8bb3aefa031

**Review by Kitzunu — APPROVED** [review]

**Comment by acidmanifesto** [comment]
> Hi, I tested your PR. Globally look fine but I found some issues with custom items:
> 
> - [ ] No icon visual in the bag
> - [ ] No visual model on the char when equipped. The weapon appear only when you attack
> - [ ] Right click to equip doesn't work for me. I need to drag and drop in my equipement panel.
> 
> I don't know if they are client issues. Otherwise no issues, new errors work fine...

Still looking into those in particular. Its listed on the todo. Some other custom items we do not have any of those issues, some only 1 or 2 of the issues, everything else is all 3. Items are functional and work. However the no visual model issue while equipped on the character i am not experiencjng with a default client. Sre you using any modded mpqs by chance? It Would help me explore and narrow down other issues if possible.

**Comment by acidmanifesto** [comment]
At the moment everything passes. The purpose of this pr was to create and use a item store while also utlizing the dbc files and correcting issues within the templates to match up with the dbc. Currently all Items is hard coded with in the source and the idea of that is absolutely insane. We utilized the dbc which is considered 100% blizz, not like or simular, but directly from blizz itself. Load times are imrpoved on server load up by a few ms in debugger.  The purpose of this pr was to utlize the dbc which it isnt currently. a quick lesson, which is copy pasta from my post in discord. DBC is a DataBase Container from what i remember in some chitchat in some of the reverse engineering rooms back then. Could be database client, as it does come from a client, didnt really care honestly, we just say dbc and leave it at that, It can improve performance. By storing data in a binary format, the DBC store can improve the performance of the game's servers. This is because the binary format is more efficient than the text format used by the SQL database. The DBC store is used to store game data that is not frequently changed, such as item definitions, creature definitions, and zone definitions. The data is consider blizzlike in the fullest. It makes it easy to update the game's data, as changes can be made to the DBC files without having to modify the game's code in most cases although certain things will require mpq modifications. Currently We have it hard coded while also, this corrects it, and no longer leaves it hard coded on a source level. Load times are imrpoved by a few ms for server started up and for character creations as well from viewed in debugger. Pr is already tested in full from a former tester and review is marked as approved.
However on the 29th i will be no longer maintaining the pr and will be closed then. Anyone is free to take it over if that comes by.

**Pull request merged**